runelite-client/src/main/java/net/runelite/client/ui/overlay/worldMap/WorldMapOverlay.java

```diff
@@ -0,0 +1,151 @@
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.ui.overlay.worldMap;
+
+import com.google.inject.Inject;
+import java.awt.Dimension;
+import java.awt.Graphics2D;
+import java.awt.Rectangle;
+import java.awt.event.MouseEvent;
+import lombok.Getter;
+import lombok.Setter;
+import net.runelite.api.Client;
+import net.runelite.api.Point;
+import net.runelite.api.coords.WorldPoint;
+import net.runelite.api.widgets.Widget;
+import static net.runelite.api.widgets.WidgetID.WORLD_MAP;
+import net.runelite.client.ui.overlay.Overlay;
+import net.runelite.client.ui.overlay.OverlayLayer;
+import net.runelite.client.ui.overlay.OverlayPosition;
+
+public abstract class WorldMapOverlay extends Overlay
+{
+	@Inject
+	WorldMapService worldMap;
+
+	@Inject
+	private Client client;
+
+	@Getter
+	private final Rectangle renderRegion = new Rectangle(0, 0, 0, 0);
+
+	@Getter
+	@Setter
+	private WorldPoint worldPoint = null;
+
+	@Getter
+	@Setter
+	private boolean snapToEdge = false;
+
+	@Inject
+	public WorldMapOverlay(Client client, WorldMapService worldMap, WorldMapOverlayMouseListener overlayMouseListener)
+	{
+		this.worldPoint = null;
+		overlayMouseListener.getOverlays().add(this);
+		setPosition(OverlayPosition.DYNAMIC);
+		setLayer(OverlayLayer.ALWAYS_ON_TOP);
+	}
+
+	@Override
+	public final Dimension render(Graphics2D graphics)
+	{
+		if (worldPoint == null)
+		{
+			this.renderRegion.setBounds(0, 0, 0, 0);
+			return null;
+		}
+
+		Widget widget = client.getWidget(WORLD_MAP, 0);
+		if (widget == null)
+		{
+			this.renderRegion.setBounds(0, 0, 0, 0);
+			return null;
+		}
+
+		Point drawPoint = worldMap.mapWorldPointToGraphicsPoint(this.worldPoint);
+
+		if (snapToEdge)
+		{
+			Rectangle mapDisplayRectangle = worldMap.getMapDisplayRectangle();
+
+			if (!mapDisplayRectangle.contains(drawPoint.getX(), drawPoint.getY()))
+			{
+				drawPoint = clipToRectangle(drawPoint, mapDisplayRectangle);
+			}
+		}
+		else
+		{
+			graphics.setClip(worldMap.getMapDisplayRectangle());
+		}
+
+		Dimension dimension = this.render(graphics, drawPoint);
+		this.renderRegion.setSize(dimension);
+		this.renderRegion.setLocation((int) (drawPoint.getX() - dimension.getWidth() / 2), (int) (drawPoint.getY() - dimension.getHeight() / 2));
```

Comment:
Will calculate incorrectly for non centered drawings. Make this render return a Rectangle instead of Dimension to fix this issue easily and cleanly.